src/backend/distributed/planner/relation_restriction_equivalence.c

```diff
@@ -1760,14 +1771,25 @@ ContainsUnionSubquery(Query *queryTree)
  * index that the partition key of the relation exists in the query. The query is
  * found in the planner info of the relation restriction.
  */
-static Index
-RelationRestrictionPartitionKeyIndex(RelationRestriction *relationRestriction)
+static Var *
+RelationRestrictionPartitionKeyIndex(Query *originalQuery,
+									 RelationRestriction *relationRestriction,
+									 Index *partitionKeyIndex)
 {
 	ListCell *targetEntryCell = NULL;
 	Index partitionKeyTargetAttrIndex = 0;
 
-	PlannerInfo *relationPlannerRoot = relationRestriction->plannerInfo;
-	Query *relationPlannerParseQuery = relationPlannerRoot->parse;
+
+	FindRteIdentityContext *findRteIdentityContext = palloc0(
+		sizeof(FindRteIdentityContext));
+	findRteIdentityContext->targetRTEIdentity = GetRTEIdentity(relationRestriction->rte);
+	FindQueryContainingRTEIdentity((Node *) originalQuery, findRteIdentityContext);
+	if (!findRteIdentityContext->found)
+	{
+		return NULL;
+	}
+
+	Query *relationPlannerParseQuery = findRteIdentityContext->query;
```

Comment:
It is weird that we use query from rteContext and targetList from parseQuery. Can't we use the former always?